### PR TITLE
[CDAP-15437] Fixed User Defined Directive (UDD) class isolation

### PR DESCRIPTION
**Pull request by nitinmotgi**
Fixed User Defined Directive (UDD) class isolation problem in DirectiveServiceHandler, upgraded GCP service versions and improved documentation around the classes that were touched.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -93,11 +93,11 @@
     <commons-lang3.version>3.5</commons-lang3.version>
     <commons-math3.version>3.6.1</commons-math3.version>
     <geogson.version>1.1.97</geogson.version>
-    <google.cloud.bigquery.version>1.36.0</google.cloud.bigquery.version>
-    <google.cloud.core.version>1.15.0</google.cloud.core.version>
+    <google.cloud.bigquery.version>1.75.0</google.cloud.bigquery.version>
```
Comment (albertshau) [inline pom.xml:96]: have we verified that this doesn't break anything?

File: wrangler-core/src/main/java/io/cdap/wrangler/registry/UserDirectiveRegistry.java [hunk 2]
```diff
@@ -112,28 +116,25 @@ public UserDirectiveRegistry(StageContext context) {
...
-          DirectiveInfo classz = new DirectiveInfo(DirectiveInfo.Scope.USER, directive);
-          namespaceRegistry.put(classz.name(), classz);
-          return classz;
-        }
+      if (context != null) {
+        directive = context.loadPluginClass(name);
       } else {
-        return namespaceRegistry.get(name);
+        if (manager != null) {
+          PluginConfigurer configurer = manager.createPluginConfigurer(namespace);
```
Comment (albertshau) [inline wrangler-core/src/main/java/io/cdap/wrangler/registry/UserDirectiveRegistry.java:125]: I believe this is a fairly expensive operation. Do you notice slowness when many directives are used? Though I guess it's probably similar to the previous implementation that loads all plugin artifacts.

File: pom.xml [hunk 3]
```diff
@@ -83,7 +83,7 @@
     <aws.sdk.version>1.11.133</aws.sdk.version>
     <bigquery.connector.hadoop2.version>0.10.2-hadoop2</bigquery.connector.hadoop2.version>
     <bouncycastle.version>1.56</bouncycastle.version>
-    <cdap.version>6.1.0-SNAPSHOT</cdap.version>
+    <cdap.version>6.0.1-SNAPSHOT</cdap.version>
```
Comment (albertshau) [inline pom.xml]: how come this is needed?

File: pom.xml (hunk 1 above)
Comment (nitinmotgi) [reply pom.xml:96]: Yes Albert, I ran my full 'root' tests with this change.

File: wrangler-core/src/main/java/io/cdap/wrangler/registry/UserDirectiveRegistry.java (hunk 2 above)
Comment (nitinmotgi) [reply wrangler-core/src/main/java/io/cdap/wrangler/registry/UserDirectiveRegistry.java:125]: I did see a bit of slowness as we don't cache it anymore. I was thinking of caching it at this point to make sure it's faster in subsequent operations, but it happens only once so I decided not to do it.

File: pom.xml (hunk 3 above)
Comment (nitinmotgi) [reply pom.xml]: Ah! will change this. When I was going the change on Saturday, the issue was 'develop' builds hadn't passed yet, so I used the release that was available to make progress. I will change this to 6.1.0-SNAPSHOT. Thanks for catching it.

**Comment by nitinmotgi** [comment]
@albertshau comments and changes in. Could you please review when u get a chance.

File: wrangler-core/src/main/java/io/cdap/wrangler/registry/UserDirectiveRegistry.java (hunk 2 above)
Comment (albertshau) [reply wrangler-core/src/main/java/io/cdap/wrangler/registry/UserDirectiveRegistry.java:125]: The correct thing would be for the service endpoint to create one PluginConfigurer, and for that same instance to be used throughout the lifetime of the call. This is a bigger refactor so I've opened https://issues.cask.co/browse/CDAP-15454 to revisit it later.

File: pom.xml (hunk 1 above)
Comment (albertshau) [reply pom.xml:96]: OK, though in the future it's better if this were done in a separate PR, as it is not related to classloading isolation.

**Review by albertshau — APPROVED** [review]
lgtm, please squash before merging

**Comment by nitinmotgi** [comment]
Squashed and merged.

**Pull request closed without merging**